### PR TITLE
Make serde lazy-load dependent modules

### PR DESCRIPTION
**Pull request by uranusjr**
Currently serde loads all possible classes on initialization, including any possible modules. This includes numpy and kubernetes, which are notoriously slow, even if they are never used.

Instead of relying on actually loading modules, this new mechanism only initializes a collection of strings at startup, and only loads relevant serialization modules when needed. This should greatly improve startup time.

### CODE REVIEW

File: airflow/serialization/serde.py [hunk 1]
```diff
@@ -270,38 +273,94 @@ def _stringify(classname: str, version: int, value: T | None) -> str:
...
+        "uint64": "airflow.serialization.serializers.numpy",
+    },
+    "pendulum": {
+        "datetime.DateTime": "airflow.serialization.serializers.datetime",
+        "tz.timezone.FixedTimezone": "airflow.serialization.serializers.timezone",
+        "tz.timezone.Timezone": "airflow.serialization.serializers.timezone",
+    },
+}
+
+_AIRFLOW_DESERIALIZERS = {
```
Comment (ashb) [inline airflow/serialization/serde.py:313]: We don't have a deserializer for Kube?

Comment (uranusjr) [reply airflow/serialization/serde.py:313]: Yes apparently, this is the case before the patch and I’m not sure why.

Comment (kaxil) [reply airflow/serialization/serde.py:313]: cc @bolkedebruin

**Comment by bolkedebruin** [comment]
Okay, I have given this some thought. The challenge is that loading certain modules is expensive. Now your current approach is to hardcode any so called `AIRFLOW_SERIALIZERS` within the `serde` module. This creates a maintenance burden as the implementation of (de)serialization happens within the implementation module - which could lead to sync issues: this already happened when you encountered that I didn't implement deserialization for a kubernetes pod yet.

The current implementation of the `serde` module already allows for string lookups. So I think it is better to do the lazy loading inside the place where the serialization happens. For example the `kubernetes` serialization could just register a string of `kubernetes.1.pod` (not the actual name) and load the module when actual (de)serialization happens. There error when trying to deserialize when the underlying module (e.g. kubernetes) isn't there is slightly different but I do not think that is a problem.

In other words I think we can decentralize the lazy loading mechanism and no change is required to the serde module. This had the benefits of mixing eager loading with lazy loading, improved maintenance and better extensibility and less code changes.

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
Please decentralize and use existing string lookup framework to lazy load.

File: airflow/serialization/serde.py (hunk 1 above)
Comment (bolkedebruin) [reply airflow/serialization/serde.py:313]: It isn't there as it is only used within DAG serialization which I havent implemented yet. But please see below, I think this 'sync' issue should be prevented by implementing the change differently.

**Comment by bolkedebruin** [comment]
Well, I think the fact that you have encountered the issue yourself proves otherwise? 

The issue is that currently you can now drop a new (de)serializer in the namespace (as an administrator). The serializers would then register themselves.With your change you cannot do that anymore and you made it static and thus require core changes when adding a (de)serializer. The plugin system was intentional and I do not think it makes sense to remove it.

If we are so afraid of hitting a performance bottleneck due to oversight during review we could `timeit` during CI/CD with a flamegraph to make it fancy or do AST parsing instead of module loading.

**Comment by uranusjr** [comment]
The “missing” numpy registrations are intentional; those are aliases and can never be identified by the `isinstance` logic.

```pycon
>>> import numpy
>>> numpy.int_
<class 'numpy.int64'>
```

This is on a 64-bit machine; on 32-bit it’s aliased to `int32`. But the point is that no object would ever report itself as an instance of `numpy.int_`, only either one of the actual int classes.

But IMO that’s beside the point either way. Missing entries is a problem no matter whether you register the types, you can equally miss adding a type to the per-module registry to adding one to the global registry. The only downside of putting all the registration in one module is when you miss registering _an entire submodule in `airflow.serialization.serializers`_, which IMO is very easy to catch (and can be automated with pre-commit if needed), and worthwhile considering how it simplifies implementing individual serializer modules in general (no need to consider import performance for each module since it’s handled by the global registry).

**Comment by uranusjr** [comment]
Continued in #30094.
